### PR TITLE
Fs subcycle cfl fix

### PR DESCRIPTION
**Pull request by ursg**
This replaces pull request #229, it was simply rebased onto current master.

The testpackage runs fine with it, but that is unsurprising, since the testpackage doesn't subcycle.

### CODE REVIEW

**Comment by ykempf** [comment]
Can you for clarity summarise at the top what this is about? It is easier than clicking through to an old/other pull request and reading through an extended thing there...

**Comment by ursg** [comment]
This pull request is a makeover of the fieldsover subcycle dt calculation, to ensure that it never oversteps and doesn't do too few steps, either.

Original found during the "summer of debugging", where field artefacts were occuring on subcycle-amount-boundaries. Fixed by yann, then fixed again in a different way by sebastian (hence the large amount of commits), then rebased onto master by urs.

**Comment by ykempf** [comment]
The derivatives of the moments needn't be recomputed in the second RK step of any subcycle, not even the first.

**Comment by ursg** [comment]
The updated testpackage shows 0 changes, except in the polar magnetosphere test and (ridiculously tiny) changes in the fieldsolver substep test:

```
--------------------------------------------------------------------------------------------
running Magnetosphere_polar_small 
Parameter data file (sw1.dat) has 1 values
--------------------------------------------------------------------------------------------
Magnetosphere_polar_small  -  Verifying dffbd6cc78031db7b0c6ae37056c0e8481528472___DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC4D_AGNER__DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC4D_AGNER against current
--------------------------------------------------------------------------------------------
------------------------------------------------------------
 ref-time     |   new-time       |  speedup                |
------------------------------------------------------------
344.6        356.3         0.967163
------------------------------------------------------------
  variable     |     absolute diff     |     relative diff | 
------------------------------------------------------------
rho_0                32.3124                 2.08945e-05    
rho_v_0                8.63193e+07                 0.000101274    
rho_v_1                6.339e+06                 0.00837116    
rho_v_2                4.5105e+07                 0.000884839    
B_0                2.91523e-12                 1.44123e-06    
B_1                2.78782e-13                 0.00369895    
B_2                5.90053e-12                 1.12096e-08    
E_0                2.37728e-07                 0.00202775    
E_1                6.30687e-06                 0.000526345    
E_2                1.42299e-07                 0.001516    
--------------------------------------------------------------------------------------------
   Distribution function diff                                                               
--------------------------------------------------------------------------------------------
INFO Reading in two files.
File names: /mnt/btrfs/work/uganse/testpackage/testpackage_data/current/Magnetosphere_polar_small/bulk.0000001.vlsv & /mnt/btrfs/work/uganse/testpackage/run_2017.02.23_13.25.53/Magnetosphere_polar_small/bulk.0000001.vlsv
NonIdenticalBlocks:      62
IdenticalBlocks:         100574
Absolute_Error:          1.81924e-10
Mean-Absolute-Error:     1.33509e-16
Max-Absolute-Error:      1.90082e-13
Absolute-log-Error:      4250.1
Mean-Absolute-log-Error: 0.00311903
--------------------------------------------------------------------------------------------
running test_fp_substeps 
--------------------------------------------------------------------------------------------
test_fp_substeps  -  Verifying dffbd6cc78031db7b0c6ae37056c0e8481528472___DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC4D_AGNER__DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC4D_AGNER against current
--------------------------------------------------------------------------------------------
------------------------------------------------------------
 ref-time     |   new-time       |  speedup                |
------------------------------------------------------------
52.88        53.47         0.988966
------------------------------------------------------------
  variable     |     absolute diff     |     relative diff | 
------------------------------------------------------------
rho_0                0                 0    
rho_v_0                0                 0    
rho_v_1                0                 0    
rho_v_2                0                 0    
B_0                5.29913e-23                 6.8043e-14    
B_1                5.32498e-23                 6.83749e-14    
B_2                5.32498e-23                 6.83749e-14    
E_0                2.12182e-19                 1.2218e-13    
E_1                2.10911e-19                 1.21448e-13    
E_2                2.11017e-19                 1.21509e-13    
--------------------------------------------------------------------------------------------
```

so I assume this is really just caused by floating point-accuracy level changes.

File: fieldsolver/ldz_main.cpp [hunk 1]
```diff
@@ -352,9 +352,9 @@ bool propagateFields(
          calculateUpwindedElectricFieldSimple(mpiGrid, sysBoundaries, localCells, RK_ORDER2_STEP1);
          
          propagateMagneticFieldSimple(mpiGrid, sysBoundaries, subcycleDt, localCells, RK_ORDER2_STEP2);
-         // If we are at the first subcycle we need to update the derivatives of the moments, 
-         // otherwise only B changed and those derivatives need to be updated.
-         calculateDerivativesSimple(mpiGrid, sysBoundaries, localCells, RK_ORDER2_STEP2, (subcycleCount==0));
+         // Here, we never need to update the moment derivatives since they will not have changed from
```
Comment (ykempf) [inline fieldsolver/ldz_main.cpp]: That is, until we interpolate moments in time as Sebastian wants us to do since last summer.

Comment (ursg) [inline fieldsolver/ldz_main.cpp]: True, but that doesn't happen yet, so this is, for now, correct.

I would suggest merging this now, if there are no other objections.

**Pull request merged**